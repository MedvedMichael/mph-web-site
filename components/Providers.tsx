import { lightTheme, darkTheme } from './themes';
import React, {useState, useEffect, createContext} from 'react'
import useDarkMode from 'use-dark-mode';
import {ThemeProvider} from 'styled-components'

const DarkThemeContext = createContext(null)

const Providers = ({ children }) => {
    const darkMode = useDarkMode(false, { storageKey: null, onChange: null })
    const theme = darkMode.value ? darkTheme : lightTheme

    const [mounted, setMounted] = useState(false)

    useEffect(() => {
        setMounted(true)
    }, [])

    

    const body =
    //@ts-ignore
    <DarkThemeContext.Provider value={darkMode}>
        <ThemeProvider theme={theme}>
            <>{children}</>
        </ThemeProvider>
    </DarkThemeContext.Provider>
    // prevents ssr flash for mismatched dark mode
    if (!mounted) {
        return <div style={{ visibility: 'hidden' }}>{body}</div>
    }

    return body
}

export default Providers

export {DarkThemeContext}