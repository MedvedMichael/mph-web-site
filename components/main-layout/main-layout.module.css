.main {
    display: flex;
    flex-direction: column;
    min-height: 100%;
}

.main-container {
    flex-grow: 1;
}