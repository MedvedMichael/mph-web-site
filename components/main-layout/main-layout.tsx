
import Head from "next/head";
import { useEffect, useState, createContext } from "react";
import styled from "styled-components";
import Footer from "../footer/footer";
import Navbar from "../navbar/navbar";

const AdminContext = createContext('')

interface MainLayout {
    title: string,
    children: JSX.Element[]
}

const MainLayout = ({ children, title }) => {
    const [isAdmin, setIsAdmin] = useState(false)

    useEffect(() => {
        const check = localStorage.getItem('isAdmin')
        if(check && check === process.env.SECRET_WORD){
            setIsAdmin(true)
        }
    })
    
    return (
        <>
            <Head>
                {/* <script src="https://kit.fontawesome.com/fc94503bd8.js" crossOrigin="anonymous"></script> */}
                <title>{title}</title>
            </Head>
            <Main>
                <AdminContext.Provider value={isAdmin ? 'admin' : ''}>
                    <Navbar />
                    <MainContainer >
                        {children}
                    </MainContainer>
                    <Footer />
                </AdminContext.Provider>
            </Main>
        </>)
}



export default MainLayout
export {AdminContext}

const Main = styled.div`
    display: flex;
    flex-direction: column;
    min-height: 65rem;
    background: ${props => props.theme.bg.primary};
    transition: ${props => props.theme.transition.bg};
`

const MainContainer = styled.main`
    padding-top: 5rem;
    display: flex;
    flex-direction: column;
    flex-grow: 1;
`