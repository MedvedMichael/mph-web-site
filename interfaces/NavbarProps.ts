export default interface NavbarProps {
    navbarState: boolean,
    handleNavbar: () => void,
    startLoading?: () => void
  }