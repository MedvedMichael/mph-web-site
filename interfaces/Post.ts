interface Comment {
    name: string,
    text: string,
    timestamp: Date,
    postId: string
}

interface Post {
    id: string,
    title: string,
    text: string,
    images?: string[],
    comments: Array<Comment>
}

export default Post
export type {Comment}