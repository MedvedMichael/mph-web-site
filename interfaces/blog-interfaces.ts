export interface Comment {
    name: string,
    text: string,
    timestamp: Date,
    postId: string
}

export interface Post {
    id: string,
    title: string,
    text: string,
    images?: string[],
    comments: Array<Comment>
}
