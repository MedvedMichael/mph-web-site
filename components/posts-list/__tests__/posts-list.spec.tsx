import PostsList from '../posts-list';
import Spinner from '../../spinner/spinner'
import { act, render, waitForElementToBeRemoved } from '@testing-library/react'
import React from 'react'
import { mount, shallow } from 'enzyme';
import { AdminContext } from '../../main-layout/main-layout';
import ShallowRenderer from 'react-test-renderer/shallow';
import styled, { ThemeProvider } from 'styled-components';
import Providers from '../../Providers';
import { Post } from '../../../interfaces/blog-interfaces';
import PostView from '../post-view';

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

// const mockHistoryPush = jest.fn();

// jest.mock("next/router", () => ({
//     useRouter() {
//         return {
//             route: "/",
//             pathname: "",
//             query: "",
//             asPath: "",
//         };
//     },
// }));

// jest.mock('next/router', () => ({
//     ...jest.requireActual('next/router'),
//     useRouter: () => ({
//         push: mockHistoryPush,
//     }),
// }));


const TestComponent = ({children, value}) => (
    <AdminContext.Provider value={value}>
        <Providers>
        {children}
        </Providers>
    </AdminContext.Provider>)

test('should render without crashing', () => {
    const wrapper = shallow(<PostsList posts={[]} />);
    // expect(wrapper.find('h1')).toHaveLength(1);
});

test('should render add button for admin', () => {
    const wrapper = mount(
        <TestComponent value='admin'>
            <PostsList posts={[]} />
        </TestComponent>)
    expect(!wrapper.find('.std-button').debug()).toBe(false)
});


test('should render message is no posts', () => {
    const wrapper = mount(
        <TestComponent value=''>
            <PostsList posts={[]} />
        </TestComponent>)
    expect(wrapper.find('#no-posts-message').get(0).props.children).toBe('There are no posts yet :(')
})

test('should render posts', async () => {
    const posts: Post[] = [
        {
            id: '1',
            title: 'Title',
            text: 'Text',
            images: [''],
            comments: []
        },
        {
            id: '2',
            title: 'Title2',
            text: 'Text2',
            images: [''],
            comments: []
        }
    ]
    
    const wrapper  = shallow(
        <TestComponent value=''>
            <PostsList posts={posts} />
        </TestComponent>)
    expect(wrapper.find(PostsList).dive().find(PostView).length).toBe(posts.length)
})

test('should push to the history on add button click', () => {
    const test = jest.fn()
    const returnFromRouter = {
        push: () => {}
    }
    const test2 = jest.spyOn(returnFromRouter, 'push')
    useRouter.mockImplementation(() => returnFromRouter)
    const wrapper = shallow(
        <TestComponent value='admin'>
            <PostsList posts={[]} />
        </TestComponent>)
    
    
    console.log(wrapper.find(PostsList).dive().debug())
    // wrapper.find('.std-button').last().simulate('click')
    // console.log(test.mock.calls)
    // expect(useRouter).toHaveBeenCalledTimes(1);
})