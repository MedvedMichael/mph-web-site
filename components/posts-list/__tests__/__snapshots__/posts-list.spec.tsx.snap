// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<styled.div>
  <styled.h1>
    Posts
  </styled.h1>
  <styled.div />
  <styled.h3
    id="no-posts-message"
  >
    There are no posts yet :(
  </styled.h3>
</styled.div>
`;
